Play 2.6.0-M5 no longer supports HTTP status code 498

### Play Version (2.5.x / etc)

2.6.0-M5

### API (Scala / Java / Neither / Both)

Scala 2.12.1

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

MacOs 10.12.4

Use `uname -a` if on Linux.

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

Paste the output from `java -version` at the command line.

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. I have a controller which creates an Action which returns a Result like this: play.api.mvc.Results.Status(498)

2. The request should return with status 498

### Actual Behavior

An exception is thrown and the request returns with status 500. This is a regression from the pre-
Akka-HTTP version of Play we used.

2017-05-04 12:05:20,594 [application-akka.actor.default-dispatcher-15] ERROR p.c.server.common.ServerResultUtils - Exception occurred while converting Result with headers [<X-API-Version>: <2.0.7>]. Calling HttpErrorHandler to get alternative Result.
java.lang.RuntimeException: Non-standard status codes cannot be created by implicit conversion. Use `StatusCodes.custom` instead.
at akka.http.scaladsl.model.StatusCode$.$anonfun$int2StatusCode$1(StatusCode.scala:28)
at scala.Option.getOrElse(Option.scala:121)
at akka.http.scaladsl.model.StatusCode$.int2StatusCode(StatusCode.scala:27)
at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$2(AkkaModelConversion.scala:183)
at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
at akka.http.scaladsl.util.FastFuture$.map$extension(FastFuture.scala:23)
at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$1(AkkaModelConversion.scala:179)
at play.core.server.common.ServerResultUtils.resultConversionWithErrorHandling(ServerResultUtils.scala:136)
at play.core.server.akkahttp.AkkaModelConversion.convertResult(AkkaModelConversion.scala:192)
at play.core.server.AkkaHttpServer.$anonfun$executeAction$3(AkkaHttpServer.scala:255)
at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1(Promise.scala:284)
at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1$adapted(Promise.scala:284)
at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
at scala.concurrent.Promise.complete(Promise.scala:49)
at scala.concurrent.Promise.complete$(Promise.scala:48)
at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:183)
at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:38)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:38)
at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
2017-05-04 12:05:20,813 [application-akka.actor.default-dispatcher-15] ERROR com.mc10inc.play.GlobalErrorHandler - Request Error
play.core.server.common.ServerResultException: Error converting Play Result for server backend
at play.core.server.common.ServerResultUtils.play$core$server$common$ServerResultUtils$$handleConversionError$1(ServerResultUtils.scala:117)
at play.core.server.common.ServerResultUtils$$anonfun$resultConversionWithErrorHandling$9.applyOrElse(ServerResultUtils.scala:136)
at play.core.server.common.ServerResultUtils$$anonfun$resultConversionWithErrorHandling$9.applyOrElse(ServerResultUtils.scala:136)
at scala.concurrent.Future.$anonfun$recoverWith$1(Future.scala:412)
at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$2(FastFuture.scala:37)
at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:46)
at akka.http.scaladsl.util.FastFuture$.transformWith$extension0(FastFuture.scala:37)
at akka.http.scaladsl.util.FastFuture$ErrorFuture.transformWith(FastFuture.scala:97)
at scala.concurrent.Future.recoverWith(Future.scala:411)
at scala.concurrent.Future.recoverWith$(Future.scala:410)
at akka.http.scaladsl.util.FastFuture$ErrorFuture.recoverWith(FastFuture.scala:88)
at play.core.server.common.ServerResultUtils.resultConversionWithErrorHandling(ServerResultUtils.scala:136)
at play.core.server.akkahttp.AkkaModelConversion.convertResult(AkkaModelConversion.scala:192)
at play.core.server.AkkaHttpServer.$anonfun$executeAction$3(AkkaHttpServer.scala:255)
at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109)
at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71)
at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1(Promise.scala:284)
at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1$adapted(Promise.scala:284)
at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
at scala.concurrent.Promise.complete(Promise.scala:49)
at scala.concurrent.Promise.complete$(Promise.scala:48)
at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:183)
at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:38)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:38)
at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.RuntimeException: Non-standard status codes cannot be created by implicit conversion. Use `StatusCodes.custom` instead.
at akka.http.scaladsl.model.StatusCode$.$anonfun$int2StatusCode$1(StatusCode.scala:28)
at scala.Option.getOrElse(Option.scala:121)
at akka.http.scaladsl.model.StatusCode$.int2StatusCode(StatusCode.scala:27)
at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$2(AkkaModelConversion.scala:183)
at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
at akka.http.scaladsl.util.FastFuture$.map$extension(FastFuture.scala:23)
at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$1(AkkaModelConversion.scala:179)
... 28 common frames omitted`
`

--- a/framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala
+++ b/framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala
@@ -13,7 +13,7 @@ import akka.stream.Materializer
 import akka.stream.scaladsl.Source
 import akka.util.ByteString
 import play.api.Logger
-import play.api.http.{ HttpChunk, HttpErrorHandler, Status, HttpEntity => PlayHttpEntity }
+import play.api.http.{ HttpChunk, HttpErrorHandler, HttpEntity => PlayHttpEntity }
 import play.api.libs.typedmap.TypedMap
 import play.api.mvc._
 import play.api.mvc.request.{ RemoteConnection, RequestTarget }
@@ -43,7 +43,7 @@ private[server] class AkkaModelConversion(
       convertRequestBody(request)
     )
 
-    //    // FIXME this is if you want to try out avoiding conversion 
+    //    // FIXME this is if you want to try out avoiding conversion
     //    (
     //      new RequestHeaderImpl(
     //        forwardedHeaderHandler.forwardedConnection(
@@ -179,8 +179,17 @@ private[server] class AkkaModelConversion(
       resultUtils.validateResult(requestHeaders, unvalidated, errorHandler).fast.map { validated: Result =>
         val convertedHeaders = convertHeaders(validated.header.headers)
         val entity = convertResultBody(requestHeaders, validated, protocol)
+        val intStatus = validated.header.status
+        val statusCode = StatusCodes.getForKey(intStatus).getOrElse {
+          val reasonPhrase = validated.header.reasonPhrase.getOrElse("")
+          if (intStatus >= 600 || intStatus < 100) {
+            StatusCodes.custom(intStatus, reasonPhrase, defaultMessage = "", isSuccess = false, allowsEntity = true)
+          } else {
+            StatusCodes.custom(intStatus, reasonPhrase)
+          }
+        }
         val response = HttpResponse(
-          status = validated.header.status,
+          status = statusCode,
           headers = convertedHeaders,
           entity = entity,
           protocol = protocol
@@ -190,7 +199,7 @@ private[server] class AkkaModelConversion(
     } {
       // Fallback response in case an exception is thrown during normal error handling
       HttpResponse(
-        status = Status.INTERNAL_SERVER_ERROR,
+        status = StatusCodes.InternalServerError,
         headers = immutable.Seq(Connection("close")),
         entity = HttpEntity.Empty,
         protocol = protocol

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/ScalaResultsHandlingSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/ScalaResultsHandlingSpec.scala
@@ -29,7 +29,7 @@ trait ScalaResultsHandlingSpec extends PlaySpecification with WsTestClient with 
 
   sequential
 
-  "scala body handling" should {
+  "scala result handling" should {
 
     def tryRequest[T](result: => Result)(block: Try[WSResponse] => T) = withServer(result) { implicit port =>
       val response = Try(await(wsUrl("/").get()))
@@ -53,6 +53,11 @@ trait ScalaResultsHandlingSpec extends PlaySpecification with WsTestClient with 
 
     "add Date header" in makeRequest(Results.Ok("Hello world")) { response =>
       response.header(DATE) must beSome
+    }
+
+    "work with non-standard HTTP response codes" in makeRequest(Result(ResponseHeader(498), HttpEntity.NoEntity)) { response =>
+      response.status must_== 498
+      response.body must beEmpty
     }
 
     "add Content-Length for strict results" in makeRequest(Results.Ok("Hello world")) { response =>

